git-branchless/src/revset/ast.rs

```diff
@@ -26,3 +27,23 @@ impl Display for Expr<'_> {
         }
     }
 }
+
+impl<'input> Expr<'input> {
+    /// Replace names in this expression with arbitrary expressions.
+    ///
+    /// Given a HashMap of names to Expr's, build a new Expr by crawling this
+    /// one and replacing any names contained in the map with the corresponding
+    /// Expr.
+    pub fn replace_names(&self, map: &HashMap<String, Expr<'input>>) -> Expr<'input> {
```

Comment:
One interesting topic is [macro hygiene](https://en.wikipedia.org/wiki/Hygienic_macro): whether names like `$1`/`$2`/etc. "leak" into the surrounding contexts and can be evaluated by code than the macro they're defined in. Can we add a test for macro hygiene? That is, if you have nested macros, can an inner one access the `$1`/`$2`/etc. from the outer one? (It probably shouldn't be able to, and I think it doesn't in this implementation.)

Reply:
OK, I've made an attempt at this, but would like your confirmation that I'm on the right path. The relevant test (and explanation) is at https://github.com/arxanas/git-branchless/pull/509/files#diff-ce578e640c3ce91a995bf4a577a8a4afcbdd9208a518df3b28998cc2c7d472e6R1161-R1201